fungible_conversion_proxy/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+    pub provider_account_id: AccountId,
+    pub owner_id: AccountId,
+}
+
+// Callback methods
+#[near_sdk::ext_contract(ext_self)]
+pub trait ExtSelfRequestProxy {
+    fn on_transfer_with_reference(
+        &self,
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+        crypto_amount: U128,
+        crypto_fee_amount: U128,
+        change: U128,
+    ) -> String;
+
+    fn ft_metadata_callback(
+        &self,
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+    ) -> Promise;
+
+    fn rate_callback(
+        &self,
+        args: PayerSuppliedArgs,
+        payer: AccountId,
+        deposit: U128,
+        payment_token_decimals: u8,
+    ) -> Promise;
+}
+
+trait FungibleTokenReceiver {
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: String, msg: String) -> Promise;
+}
+
+#[near_bindgen]
+impl FungibleTokenReceiver for FungibleConversionProxy {
+    /// This is the function that will be called by the fungible token contract's `ft_transfer_call` function.
+    /// We use the `msg` field to obtain the arguments supplied by the caller specifying the intended payment.
+    /// `msg` should be a string in JSON format containing all the fields in `PayerSuppliedArgs`.
+    /// Eg. msg = {"payment_reference":"abc7c8bb1234fd12","to":"dummy.payee.near","amount":"1000000","currency":"USD","token_address":"a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48.factory.bridge.near","fee_address":"fee.requestfinance.near","fee_amount":"200","max_rate_timespan":"0"}
+    ///
+    /// For more information on the fungible token standard, see https://nomicon.io/Standards/Tokens/FungibleToken/Core
+    ///
+    #[payable]
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: String, msg: String) -> Promise {
+        let args: PayerSuppliedArgs = serde_json::from_str(&msg).expect("Incorrect msg format");
+        self.transfer_with_reference(args, sender_id, U128::from(amount.parse::<u128>().unwrap()))
+    }
+}
+
+#[near_bindgen]
+impl FungibleConversionProxy {
+    /// Main function for this contract, transfers fungible tokens to a payment address (to) with a payment reference, as well as a fee.
+    /// The `amount` is denominated in `currency` with 2 decimals.
+    ///
+    /// Due to the way NEAR defines the fungible token standard, this function should NOT be called directly. Instead, the
+    /// `ft_transfer_call` function in the contract of the fungible token being used for payment should be called with the
+    /// `msg` argument containing the arguments for this function as a JSON-serialized string. `ft_on_transfer` (defined
+    /// in this contract) is called by the token contract and deserializes the arguments and calls this function.
+    ///
+    /// See https://nomicon.io/Standards/Tokens/FungibleToken/Core for more information on how NEAR handles
+    /// sending fungible tokens to be used by a contract function.
+    ///
+    #[payable]
```

Comment:
```suggestion
```
Isn't payable for paying the method with NEAR tokens?

Reply:
So since this function performs two transfers (main payment and fee payment) does that mean it requires 2 yoctoNEAR?

Reply:
I think that the minimum deposit applies to `ft_transfer` but not to `ft_on_transfer`, I can see an example from the Near foundation here: https://github.com/near-examples/FT/blob/334b32d65313f8c78d614e7cb4ba629af3af0fd9/test-contract-defi/src/lib.rs#L37 that is not payable.
If this is a correct analogy, none of these methods should be `payable`.

I agree that this is not related to gas saving but more for future maintainers to understand the methods.